app/src/main/java/com/github/swent/echo/data/repository/RepositoryImpl.kt

```diff
@@ -1,65 +1,209 @@
-        return remoteDataSource.createEvent(event)
+        if (isOffline()) {
+            throw RepositoryStoreWhileNoInternetException("Event")
+        }
+        val newEventId = remoteDataSource.createEvent(event)
+        localDataSource.setEvent(event.copy(eventId = newEventId))
+        return newEventId
     }
 
     override suspend fun setEvent(event: Event) {
-        return remoteDataSource.setEvent(event)
+        if (isOffline()) {
+            throw RepositoryStoreWhileNoInternetException("Event")
+        }
+        remoteDataSource.setEvent(event)
+        localDataSource.setEvent(event)
     }
 
     override suspend fun getAllEvents(): List<Event> {
-        return remoteDataSource.getAllEvents()
+        if (isOffline() || notExpired(events_last_cached_all, EVENT_CACHE_TTL)) {
+            return localDataSource.getAllEvents(FETCH_ALL)
+        }
+        val nonExpired = localDataSource.getAllEvents(EVENT_CACHE_TTL)
+        val nonExpiredIds = nonExpired.map { it.eventId }
+        val remoteUpdate = remoteDataSource.getEventsNotIn(nonExpiredIds)
+        localDataSource.deleteEventsNotIn(nonExpiredIds)
+        localDataSource.setEvents(remoteUpdate)
+        events_last_cached_all = currentTimeStamp()
+        return nonExpired + remoteUpdate
     }
 
     override suspend fun joinEvent(userId: String, event: Event): Boolean {
-        return remoteDataSource.joinEvent(userId, event.eventId)
+        if (isOffline()) {
+            throw RepositoryStoreWhileNoInternetException("Event Join")
+        }
+        val remoteResult = remoteDataSource.joinEvent(userId, event.eventId)
+        localDataSource.joinEvent(userId, event.eventId)
+        return remoteResult
     }
 
     override suspend fun leaveEvent(userId: String, event: Event): Boolean {
-        return remoteDataSource.leaveEvent(userId, event.eventId)
+        if (isOffline()) {
+            throw RepositoryStoreWhileNoInternetException("Event Join")
+        }
+        val remoteResult = remoteDataSource.leaveEvent(userId, event.eventId)
+        localDataSource.leaveEvent(userId, event.eventId)
+        return remoteResult
     }
 
     override suspend fun getJoinedEvents(userId: String): List<Event> {
-        return remoteDataSource.getJoinedEvents(userId)
+        if (
+            isOffline() ||
+                notExpired(events_last_cached_joined.getOrDefault(userId, 0), EVENT_CACHE_TTL)
+        ) {
+            return localDataSource.getJoinedEvents(userId, FETCH_ALL)
+        }
+        val nonExpired = localDataSource.getJoinedEvents(userId, EVENT_CACHE_TTL)
+        val nonExpiredIds = nonExpired.map { it.eventId }
+        val remoteUpdate = remoteDataSource.getJoinedEventsNotIn(userId, nonExpiredIds)
+        localDataSource.leaveEventsNotIn(userId, nonExpiredIds)
+        localDataSource.joinEvents(userId, remoteUpdate.map { it.eventId })
+        events_last_cached_joined.put(userId, currentTimeStamp())
```

Comment:
Nitpicking: could be converted to an assignment `events_last_cached_joined[userId] = currentTimeStamp()`.

Reply:
yes, you're right, I've considered that; however personally think calling .put is more explicit for the actual action performed, so that's why I prefer it.